rewrite-java/src/main/java/org/openrewrite/java/RenameVariable.java

```diff
@@ -38,35 +43,91 @@ public RenameVariable(J.VariableDeclarations.NamedVariable variable, String toNa
-            if (!isConstant && multiVariable.getVariables().contains(variable)) {
-                if (multiVariable.hasModifier(J.Modifier.Type.Static) && multiVariable.hasModifier(J.Modifier.Type.Final)) {
-                    isConstant = true;
+        public Statement visitStatement(Statement statement, P p) {
+            if (getCursor().getParent() != null
+                    && !(getCursor().getParent().getValue() instanceof J.CompilationUnit)
+                    && !(getCursor().getParent().getValue() instanceof J.ClassDeclaration)) {
+
+                // The current statement is out of scope of the previous variable declaration.
+                if (currentBlock != null && !getCursor().isScopeInPath(currentBlock)
+                        && currentNameScope.size() > 0 && !getCursor().isScopeInPath(currentNameScope.peek())) {
+                    currentNameScope.pop();
                 }
             }
-            return super.visitVariableDeclarations(multiVariable, p);
+            return super.visitStatement(statement, p);
+        }
+
+        @Override
+        public J.VariableDeclarations.NamedVariable visitVariable(J.VariableDeclarations.NamedVariable namedVariable, P p) {
+            if (namedVariable.getSimpleName().equals(variable.getSimpleName())) {
+                Cursor parentScope = getCursorToParentScope(getCursor());
+                // The target variable was found and is not scoped to the class declaration.
+                if ((currentNameScope.size() == 0 && namedVariable.equals(variable))
+                        // A variable has been declared in a new name scope.
+                        || (currentNameScope.size() > 0 && !parentScope.getValue().equals(currentNameScope.peek())
+                            && getCursor().isScopeInPath(currentNameScope.peek()))) {
+
+                    currentBlock = parentScope.getValue();
+                    currentNameScope.add(parentScope.getValue());
+                }
+            }
+            return super.visitVariable(namedVariable, p);
+        }
+
+        /**
+         * Returns either the current block or a J.Type that may create a reference to a variable.
+         * I.E. for(int target = 0; target < N; target++) creates a new name scope for `target`.
+         * The name scope in the next J.Block `{}` cannot create new variables with the name `target`.
+         */
+        private Cursor getCursorToParentScope(Cursor cursor) {
+            return cursor.dropParentUntil(t -> t instanceof J.Block
+                    || t instanceof J.ForLoop
+                    || t instanceof J.Case
+                    || t instanceof J.Try
+                    || t instanceof J.Try.Resource
+                    || t instanceof J.Try.Catch
+                    || t instanceof J.MultiCatch
+                    || t instanceof J.MethodDeclaration
+                    || t instanceof J.Lambda);
```

Comment:
I recommend looking through all the J.* classes, I think we are missing a couple here: ForEach, While, etc.

Reply:
Ah, Jon mentioned ForEach, but I forgot to add it when I swept through the code.
From what I tested the rest of the J.* types don't create a new name scope and reference an existing variable in a parent scope. Case statements were an exception.